consensus/istanbul/backend/internal/enodes/val_enode_db.go

```diff
@@ -0,0 +1,312 @@
+	return newPersistentDB(path)
+}
+
+// newMemoryDB creates a new in-memory node database without a persistent backend.
+func newMemoryDB() (*ValidatorEnodeDB, error) {
+	db, err := leveldb.Open(storage.NewMemStorage(), nil)
+	if err != nil {
+		return nil, err
+	}
+	return &ValidatorEnodeDB{
+		db: db,
+	}, nil
+}
+
+// newPersistentNodeDB creates/opens a leveldb backed persistent node database,
+// also flushing its contents in case of a version mismatch.
+func newPersistentDB(path string) (*ValidatorEnodeDB, error) {
+	opts := &opt.Options{OpenFilesCacheCapacity: 5}
+	db, err := leveldb.OpenFile(path, opts)
+	if _, iscorrupted := err.(*lvlerrors.ErrCorrupted); iscorrupted {
+		db, err = leveldb.RecoverFile(path, nil)
+	}
+	if err != nil {
+		return nil, err
+	}
+	// The nodes contained in the cache correspond to a certain protocol version.
+	// Flush all nodes if the version doesn't match.
+	currentVer := make([]byte, binary.MaxVarintLen64)
+	currentVer = currentVer[:binary.PutVarint(currentVer, int64(dbVersion))]
+
+	blob, err := db.Get([]byte(dbVersionKey), nil)
+	switch err {
+	case leveldb.ErrNotFound:
+		// Version not found (i.e. empty cache), insert it
+		if err := db.Put([]byte(dbVersionKey), currentVer, nil); err != nil {
+			db.Close()
+			return nil, err
+		}
+
+	case nil:
+		// Version present, flush if different
+		if !bytes.Equal(blob, currentVer) {
+			db.Close()
+			if err = os.RemoveAll(path); err != nil {
+				return nil, err
+			}
+			return newPersistentDB(path)
+		}
+	}
+	return &ValidatorEnodeDB{db: db}, nil
+}
+
+// Close flushes and closes the database files.
+func (vet *ValidatorEnodeDB) Close() error {
+	return vet.db.Close()
+}
+
+func (vet *ValidatorEnodeDB) String() string {
+	var b strings.Builder
+	b.WriteString("ValEnodeTable:")
+
+	err := vet.iterateOverAddressEntries(func(address common.Address, entry *addressEntry) error {
+		fmt.Fprintf(&b, " [%s => %s]", address.String(), entry.String())
+		return nil
+	})
+
+	if err != nil {
+		log.Error("ValidatorEnodeDB.String error", "err", err)
+	}
+
+	return b.String()
+}
+
+// GetEnodeURLFromAddress will return the enodeURL for an address if it's known
+func (vet *ValidatorEnodeDB) GetEnodeURLFromAddress(address common.Address) (string, error) {
+	entry, err := vet.getAddressEntry(address)
+	if err != nil {
+		return "", err
+	}
+	return entry.enodeURL, nil
+}
+
+// GetAddressFromEnodeURL will return the address for an enodeURL if it's known
+func (vet *ValidatorEnodeDB) GetAddressFromEnodeURL(enodeURL string) (common.Address, error) {
+	rawEntry, err := vet.db.Get(enodeURLKey(enodeURL), nil)
+	if err != nil {
+		return common.ZeroAddress, err
+	}
+	return common.BytesToAddress(rawEntry), nil
+}
+
+// Upsert will update or insert a validator enode entry; given that the existing entry
+// is older (determined by view parameter) that the new one
+func (vet *ValidatorEnodeDB) Upsert(remoteAddress common.Address, enodeURL string, view *istanbul.View) (string, error) {
```

Comment:
Should this deal with the case if the entry is not changed?

Reply:
done!